Deployment Module Noseover Triggers Pyro Charges

AarC10 opened the pull request:
This closes #252

# Description
Have a noseover detection alert trigger the pyro charges on the deployment module where any closed loop on the sense will trigger charges. This is just basic functionality to test the boards when they arrive and charges are not expected to do anything during the OTIS flight. Log when we receive a noseover notification and when we deploy charges. A future change should support the settings subsystem in Zephyr where we can set what delay we want for deployments.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Can't really be tested until we get hardware (minus wiring a bunch on a nucleo). Will test when we get the hardware and submit bug fixes when appropriate.

# Checklist:
- [x] New functionality is documented in the necessary spots (i.e new functions documented in the header)
- [x] Unit tests cover any new functionality or edge cases that the PR was meant to resolve (if applicable) 
- [x] The CI checks are passing
- [x] I reviewed my own code in the GitHub diff and am sure that each change is intentional
- [x] I feel comfortable about this code flying in a rocket

ZZwatermelon commented on app/backplane/deployment_module/src/c_pyro_control_observer.cpp:
```diff
@@ -0,0 +1,50 @@
+        case NAlerts::NOSEOVER:
+            LOG_INF("Noseover detected. Deploying charges.");
+            logTimestamp();
+            logFile.Write(" - Noseover detected. Deploying charges.\n", 40);
+            // TODO: Settings library for handling deployment timing
```
So obv we don't actually want to deploy all our charges all at once, what is the plan for how to set the delay?
Since this isn't final dep mod, could we just hardcode a 1s delay after the second charge?
(Side not how many charges does this rev of dep mod support?)

AarC10 replied:
In this current state, its just for testing basic functionality when we get the board. Also, this was the plan
https://docs.zephyrproject.org/latest/services/storage/settings/index.html

cowsed commented on app/backplane/deployment_module/include/c_pyro_control_observer.h:
```diff
@@ -0,0 +1,64 @@
+        CGpio(GPIO_DT_SPEC_GET(DT_ALIAS(gpio3), gpios))
+    };
+
+    CFile logFile{"deployment_log.txt", CFile::READ_WRITE_FLAG | CFile::CREATE_FLAG | CFile::APPEND_FLAG};
```
Any reason for not using CFlightLog, would get you timestamps for free so you don't have to add them manually like you do in this .cpp. 

Im seeing that you use an rtc time which it doesnt currently support but it has overloads for custom integerr timestamps, it wouldnt be a big jump to add an overload for custom string timestamp. imo that file is for exactly this purpose of needing to write timestamped log messages. CFile does the trick but it feels like its no necessarily the best tool for the job

cowsed replied:
Also, we should probably make it so this filename is modified so it doesn't overwrite itself

cowsed replied:
Also also, does that filename work? usually i have to do something like `/lfs/deployment_log.txt` or else zephyr doesn't know what mount point to put it at

AarC10 replied:
Pretty sure we just append the log. Also, yea forgot the mp. Using flight log now

cowsed replied:
word

cowsed commented on app/backplane/deployment_module/src/c_pyro_control_observer.cpp:
```diff
@@ -0,0 +1,50 @@
+                    logFile.Write(pyroCountStr, 1);
+                    logFile.Write(" - Deployed pyro ", 17);
+
+                    ctrl.SetPin(1);
```
kinda paranoia but could you put the setting pin before the IO operations. If we're worried about minimizing latency of charge deployments so we don't start falling too fast before charges go off, putting the GPIO before the operations that hit uart and flash could save a bit of time

cowsed reviewed (approved):
![image](https://github.com/user-attachments/assets/0f3bbab3-4045-434b-8f39-636ae32aea12)
we love to see it

ZZwatermelon commented on app/backplane/deployment_module/src/c_pyro_control_observer.cpp:
```diff
@@ -0,0 +1,39 @@
+
+            for (auto& [sense, ctrl] : pyroPairs) {
+                if (sense.GetPin() == 1) {
+                    ctrl.SetPin(1);
```
Do we just leave the ctrl pin high after we deploy a charge?
I'm not 100% sure about the reasoning, but I'm pretty sure most Flight computers don't do that, they only leave them high for a few seconds.

ZZwatermelon replied:
or is this a 'planned for later' typa thing

AarC10 replied:
Few second delay now. Might as well.

AarC10 commented:
@cowsed Sorry. You might want to re-review since I just added Stephen's above request and we probably shouldn't leave it high all the time even if the circuit gets broken. Also, added syncing the LEDs with the charges for more debuggability.

cowsed reviewed (approved):
looks good, good catch stephen

AarC10 commented on app/backplane/deployment_module/src/c_pyro_control_observer.cpp:
```diff
@@ -0,0 +1,58 @@
+    for (auto& [sense, ctrl, led] : pyroTrios) {
+        ctrl.SetPin(0);
+        led.SetPin(0);
+        LOG_INF("Disabled charge %d", pyroCount);
```
No you're so right lol. Fixed
